File: implement-shell-tools/wc/wc.js
const fs = require("fs");

const args = process.argv.slice(2);

let flagL = false;
let flagW = false;
let flagC = false;
const files = [];

for (const arg of args) {
  if (arg === "-l") flagL = true;
  else if (arg === "-w") flagW = true;
  else if (arg === "-c") flagC = true;
  else files.push(arg);
}

const noFlags = !flagL && !flagW && !flagC;

let totalLines = 0;
let totalWords = 0;
let totalBytes = 0;

for (const file of files) {
  const content = fs.readFileSync(file, "utf-8");
  const bytes = fs.statSync(file).size;
  const lines = content.split("\n").length - 1;
  const words = content.split(/\s+/).filter(w => w !== "").length;

  totalLines += lines;
  totalWords += words;
  totalBytes += bytes;

  const parts = [];
  if (flagL || noFlags) parts.push(String(lines).padStart(8));
  if (flagW || noFlags) parts.push(String(words).padStart(8));
  if (flagC || noFlags) parts.push(String(bytes).padStart(8));
  parts.push(" " + file);

  console.log(parts.join(""));
}

if (files.length > 1) {
  const parts = [];
  if (flagL || noFlags) parts.push(String(totalLines).padStart(8));
  if (flagW || noFlags) parts.push(String(totalWords).padStart(8));
  if (flagC || noFlags) parts.push(String(totalBytes).padStart(8));
  parts.push(" total");

  console.log(parts.join(""));
}

File: implement-shell-tools/wc/wc.py
import sys
import os

args = sys.argv[1:]

flag_l = False
flag_w = False
flag_c = False
files = []

for arg in args:
    if arg == "-l":
        flag_l = True
    elif arg == "-w":
        flag_w = True
    elif arg == "-c":
        flag_c = True
    else:
        files.append(arg)

no_flags = not flag_l and not flag_w and not flag_c

total_lines = 0
total_words = 0
total_bytes = 0

for file in files:
    with open(file, "r") as f:
        content = f.read()
    byte_count = os.path.getsize(file)
    line_count = content.count("\n")
    word_count = len(content.split())

    total_lines += line_count
    total_words += word_count
    total_bytes += byte_count

    parts = []
    if flag_l or no_flags:
        parts.append(f"{line_count:>8}")
    if flag_w or no_flags:
        parts.append(f"{word_count:>8}")
    if flag_c or no_flags:
        parts.append(f"{byte_count:>8}")
    parts.append(f" {file}")
    print("".join(parts))

if len(files) > 1:
    parts = []
    if flag_l or no_flags:
        parts.append(f"{total_lines:>8}")
    if flag_w or no_flags:
        parts.append(f"{total_words:>8}")
    if flag_c or no_flags:
        parts.append(f"{total_bytes:>8}")
    parts.append(" total")
    print("".join(parts))